zerver/lib/idempotent_request.py

```diff
@@ -0,0 +1,180 @@
+    the corresponding realm, user, idempotency_key in the idempotency table.
+
+    2- Ensure 1.idempotency and 2.that only one request does the same work at a time.
+
+    Returns: id of the new inserted row.
+
+    First SELECT: selects the corresponding row from idempotency table
+    and locks the row, but abort if another request is holding the lock (using FOR UPDATE NOWAIT)
+
+    We have two CTEs:
+
+    1.insertion_cte:
+    Inserts a new row returning the new db-generated id of that inserted row.
+    Note: The INSERT statement is not conditional as this query runs ONLy for
+    new work (e.g. non-duplicate message) except in the case of a concurrent request
+    which is handled correctly by FOR UPDATE NOWAIT.
+
+    2.update_cte:
+    References insertion_cte to SET insertion_cte.identifier to the
+    id of the new inserted row (insertion_cte.new_id).
+    Note: nothing actually updates if insertion_cte returns no rows,
+    but insertion_cte should always inserts and returns the id.
+    """
+
+    with connection.connection.cursor(cursor_factory=DictCursor) as cursor:
+        query = SQL(
+            """
+            SELECT identifier FROM {idempotency_table}
+            WHERE realm_id = %(realm_id)s
+            AND user_id = %(user_id)s
+            AND idempotency_key = %(idempotency_key_value)s
+            FOR UPDATE NOWAIT;
+
+            WITH insertion_cte AS(
+                INSERT INTO {dest_table} ({cols})
+                SELECT {values_placeholders}
+                RETURNING {pk} AS new_id
+            ),
+            update_cte AS(
+                UPDATE {idempotency_table}
+                SET identifier = insertion_cte.new_id FROM insertion_cte
+                WHERE realm_id = %(realm_id)s
+                AND user_id = %(user_id)s
+                AND idempotency_key = %(idempotency_key_value)s
+            )
+            SELECT new_id FROM insertion_cte AS new_id
+        """
+        ).format(
+            dest_table=Identifier(destination_model._meta.db_table),
+            idempotency_table=Identifier(IdempotentRequest._meta.db_table),
+            cols=SQL(", ").join(map(Identifier, field_names)),
+            values_placeholders=SQL(", ").join(map(Placeholder, field_names)),
+            pk=Identifier(destination_model._meta.pk.name),
+        )
+        cursor.execute(
+            query,
+            {
+                **column_to_value,
+                "idempotency_key_value": idempotency_key,
+                "realm_id": realm.id,
+                "user_id": user.id,
+            },
+        )
+        result = cursor.fetchall()
+
+        # We should only get a single result row.
+        assert len(result) == 1
+        return result[0]
+
+
+def idempotent_request_transaction(
+    durable: bool = False,
+    savepoint: bool = True,
+    *,
+    cached_response_serializer: Callable[[Any], dict[str, Any]],
+    cached_response_deserializer: Callable[[str], Any],
```

Comment:
The fact that these are not inverses of each other is a smell.  Specifically, the `cached_response` database column is a JSONB column, so we _know_ it's not just going to be random string.  Every `cached_response_deserializer` is going to need to call `orjson.loads`.  We can, and should, do that JSON decoding in `idempotent_request_transaction` for them, and pass the deserializer a `dict[str, Any]` -- which will make the types of these two functions nicely inverses of each other, as they should be.